### ISSUE TITLE
nn.LSTM internal dropout causes output collapse when model weights are trained on MPS and moved to CPU

### ISSUE THREAD

**Issue by christiansegercrantz** [pytorch/pytorch#173640]
When training an `nn.LSTM` with `dropout > 0` on MPS backend and then moving the model to CPU for inference, the output logits collapse significantly (~19x reduction in standard deviation). The same model with `dropout=0.0` works correctly.

**Key observations:**
- MPS inference: logit std ~1.8 (correct)
- CPU inference: logit std ~0.1 (collapsed, ~19x reduction)
- Setting `dropout=0.0` in LSTM: both MPS and CPU produce identical results

The issue is specifically with the LSTM's **internal dropout** (`dropout` parameter between layers), not `nn.Dropout` applied before or after the LSTM.

**Expected behavior:** Model outputs should be equivalent on MPS and CPU (accounting for minor floating-point differences).

**Actual behavior:** When LSTM has `dropout > 0`, outputs collapse by ~19x when transferred from MPS to CPU.

## Minimal reproducer

```python
"""Minimal reproducer: MPS LSTM internal dropout causes logit collapse on CPU."""
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, TensorDataset

torch.manual_seed(42)


class LSTMModel(nn.Module):
    def __init__(self, dropout: float = 0.2) -> None:
        super().__init__()
        self.embedding = nn.Embedding(1000, 128, padding_idx=0)
        self.lstm = nn.LSTM(128, 64, num_layers=4, batch_first=True, bidirectional=True, dropout=dropout)
        self.fc = nn.Linear(64, 2)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        _, (hidden, _) = self.lstm(self.embedding(x))
        return self.fc(hidden[-1])


def main() -> None:
    if not torch.backends.mps.is_available():
        print("ERROR: MPS not available")
        return

    # Create synthetic data
    x = torch.cat([torch.randint(1, 500, (700, 20)), torch.randint(500, 1000, (300, 20))])
    y = torch.cat([torch.zeros(700), torch.ones(300)]).long()
    loader = DataLoader(TensorDataset(x, y), batch_size=64, shuffle=True)

    print("BUG: nn.LSTM dropout causes MPS→CPU logit collapse")

    for dropout in [0.2, 0.0]:
        torch.manual_seed(42)
        model = LSTMModel(dropout=dropout)
        model.to("mps")
        opt = torch.optim.AdamW(model.parameters(), lr=1e-4)

        # Train on MPS
        for _ in range(10):
            model.train()
            for xb, yb in loader:
                opt.zero_grad()
                F.cross_entropy(model(xb.to("mps")), yb.to("mps")).backward()
                opt.step()

        # Compare MPS vs CPU inference
        model.eval()
        with torch.no_grad():
            std_mps = torch.cat([model(xb.to("mps")).cpu() for xb, _ in loader]).std().item()
            model.to("cpu")
            std_cpu = torch.cat([model(xb) for xb, _ in loader]).std().item()

        ratio = std_mps / std_cpu if std_cpu > 0.01 else float("inf")
        status = "BUG" if ratio > 2 else "OK"
        print(f"LSTM dropout={dropout}: MPS std={std_mps:.3f}, CPU std={std_cpu:.3f}, Ratio={ratio:.1f}x [{status}]")


if __name__ == "__main__":
    main()
```

## Observed output

```
BUG: nn.LSTM dropout causes MPS→CPU logit collapse
LSTM dropout=0.2: MPS std=1.811, CPU std=0.094, Ratio=19.3x [BUG]
LSTM dropout=0.0: MPS std=1.997, CPU std=1.997, Ratio=1.0x [OK]
```

## Expected output

Both models should show ratio ~1.0x (MPS and CPU outputs should be equivalent).

## Workarounds

1. Train on CPU (`PYTORCH_ENABLE_MPS_FALLBACK=1` or force CPU device)
2. Set `dropout=0.0` in LSTM (loses regularization benefit)
3. Keep inference on MPS (not always practical for production deployment)

### Versions

Collecting environment information...
PyTorch version: 2.9.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.2 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: version 4.1.2
Libc version: N/A

Python version: 3.13.9 (main, Nov 19 2025, 23:39:32) [Clang 21.1.4 ] (64-bit runtime)
Python platform: macOS-26.2-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] numpy==2.3.3
[pip3] pytorch-ignite==0.5.2
[pip3] torch==2.9.1
[pip3] torcheval==0.0.7
[conda] Could not collect

cc @mikaylagawarecki @malfet @aditvenk @kulinseth @DenisVieriu97 @jhavukainen

**Comment by christiansegercrantz** [comment]
As additional information, you can temporarily work around the issue by "exploding" the last like this:
```python
        self.lstm1 = nn.LSTM(
            embed_dim, hidden_dim, num_layers=1, batch_first=True, bidirectional=True
        )
        self.lstm_drop1 = nn.Dropout(lstm_dropout)
        self.lstm2 = nn.LSTM(
            lstm_output_dim, hidden_dim, num_layers=1, batch_first=True, bidirectional=True
        )
        self.lstm_drop2 = nn.Dropout(lstm_dropout)
        self.lstm3 = nn.LSTM(
            lstm_output_dim, hidden_dim, num_layers=1, batch_first=True, bidirectional=True
        )
        ...
```

**Comment by christiansegercrantz** [comment]
Am I reading correctly that this has been fixed and merged in https://github.com/pytorch/pytorch/pull/180873 ? If so, we can close this issue as well. 

Is it already in an available non-preview version of PyTorch?

**Comment by Isalia20** [comment]
Yes, if you install the nightly version and try, I get these:
```
BUG: nn.LSTM dropout causes MPS→CPU logit collapse
LSTM dropout=0.2: MPS std=0.173, CPU std=0.173, Ratio=1.0x [OK]
LSTM dropout=0.0: MPS std=1.968, CPU std=1.968, Ratio=1.0x [OK]
```

Let me know if it's no longer issue on your end as well and I'll close the issue. @christiansegercrantz
